src/audio/drc/drc.c

```diff
@@ -0,0 +1,849 @@
+	}
+
+	state->detector_average = 0;
+	state->compressor_gain = DRC_ONE_Q30;
+
+	state->last_pre_delay_frames = DRC_DEFAULT_PRE_DELAY_FRAMES;
+	state->pre_delay_read_index = 0;
+	state->pre_delay_write_index = DRC_DEFAULT_PRE_DELAY_FRAMES;
+
+	state->max_attack_compression_diff_db = INT32_MIN;
+	state->ratio = DRC_MINUS_ONE_Q30;
+	state->slope = DRC_MINUS_ONE_Q22;
+
+	state->linear_threshold = DRC_MINUS_ONE_Q30;
+	state->db_threshold = DRC_MINUS_ONE_Q24;
+
+	state->db_knee = DRC_MINUS_ONE_Q24;
+	state->knee_threshold = DRC_MINUS_ONE_Q30;
+	state->ratio_base = DRC_MINUS_ONE_Q30;
+
+	state->K = DRC_MINUS_ONE_Q12;
+}
+
+static inline int drc_init_pre_delay_buffers(struct drc_state *state,
+					     size_t sample_bytes,
+					     int32_t channels)
+{
+	int i;
+
+	/* Allocate pre-delay (lookahead) buffers */
+	for (i = 0; i < channels; ++i) {
+		state->pre_delay_buffers[i] =
+			rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				sample_bytes * DRC_MAX_PRE_DELAY_FRAMES);
+		if (!state->pre_delay_buffers[i])
+			return -ENOMEM;
+		memset(state->pre_delay_buffers[i], 0, sample_bytes * DRC_MAX_PRE_DELAY_FRAMES);
+	}
+
+	return 0;
+}
+
+static inline void set_pre_delay_time(struct drc_state *state,
+				      int32_t pre_delay_time,
+				      int32_t rate)
+{
+	uint32_t pre_delay_frames;
+
+	/* Re-configure look-ahead section pre-delay if delay time has
+	 * changed. */
+	pre_delay_frames = pre_delay_time * rate;
+	pre_delay_frames = min(pre_delay_frames, DRC_MAX_PRE_DELAY_FRAMES - 1);
+
+	/* Make pre_delay_frames multiplies of DIVISION_FRAMES. This way we
+	 * won't split a division of samples into two blocks of memory, so it is
+	 * easier to process. This may make the actual delay time slightly less
+	 * than the specified value, but the difference is less than 1ms. */
+	pre_delay_frames &= ~DRC_DIVISION_FRAMES_MASK;
+
+	/* We need at least one division buffer, so the incoming data won't
+	 * overwrite the output data */
+	pre_delay_frames = max(pre_delay_frames, DRC_DIVISION_FRAMES);
+
+	if (state->last_pre_delay_frames != (int32_t)pre_delay_frames) {
+		state->last_pre_delay_frames = (int32_t)pre_delay_frames;
+		state->pre_delay_read_index = 0;
+		state->pre_delay_write_index = (int32_t)pre_delay_frames;
+	}
+}
+
+/* Input: Q12.20 -> Output: Q12.20 */
+static inline int32_t knee_expf(int32_t x)
+{
+#ifdef DRC_SLOW_MATHEMATICS
+	return Q_CONVERT_FLOAT(expf(Q_CONVERT_QTOF(x, 20)), 20);
+#else
+	/* exp(x) = decibels_to_linear(20*log10(e)*x) */
+	return db2lin_fixed(8.685889638065044f * x);
+#endif
+}
+
+/* Input: Q12.20, Q20.12 -> Output: Q12.20 */
+static int32_t knee_curve(struct drc_state *state, int32_t x, int32_t k)
+{
+	int64_t num;
+	/* Linear up to threshold */
+	if (x < state->linear_threshold)
+		return x;
+
+	/* Q12.20 -> Q24.40 */
+	num = Q_SHIFT_LEFT((int64_t)(Q_CONVERT_FLOAT(1, 20) - knee_expf(Q_MULTSR_32X32((int64_t)(-k), (x - dk->linear_threshold), 12, 20, 20))), 20, 32);
+	return state->linear_threshold + (int32_t)(num / k);
```

Comment:
out of interest, will k ever be a power of 2 ? (since shifting is always cheaper than 64bit division)

Reply:
Also if its harded maybe we shouldnt be passing it around as an argument

Reply:
K is not hardcoded to 5, K is the pre-calculated value determined by the compression ratio.
Then I think the division is unavoidable.

Reply:
Is this pre-calculated based on coefficients ? Can the coefficients were generated for a k = power of 2, then it turn into a shift ?